### ISSUE TITLE
cosine_distance returns (1-sim)/2 locally but 1-sim when federated to DuckDB — same query, scores 2x apart

### ISSUE THREAD

**Issue by claudespice** [spiceai/spiceai#11263]
Found during a data-correctness audit of trunk.

`crates/runtime-datafusion-udfs/src/cosine_distance.rs:225-228` computes `(1.0 - similarity) / 2.0` (range [0,1]), while the UDF is allowed to federate to DuckDB where it unparses to `array_cosine_distance` = `1 - sim` (range [0,2]) — `crates/runtime/src/datafusion/udf.rs:668`, `crates/data-connectors/connector-duckdb/src/lib.rs:80-81`.

Consequences:
- `SELECT cosine_distance(a, b) FROM t` returns values 2x apart depending on whether the plan federates.
- `vector_search`/`/v1/search` compute score = `1 - cosine_distance`, documented as "cosine similarity" (`crates/runtime/src/embeddings/udtf.rs:180,1066-1074`) — with the local formula the actual score is `(1+sim)/2`, so orthogonal vectors score 0.5, while the DuckDB search-index path (`crates/search/src/index/duckdb/sql.rs:341`) yields the true `sim`. User score thresholds (e.g. `score > 0.7`) silently select different rows per execution path. Ranking within one path is unaffected (monotonic).
- Related: no zero-vector guard — all-zero embeddings produce NaN, which sorts above all numbers in `ORDER BY _score DESC` (`cosine_distance.rs:225`), surfacing failed/zero embeddings as top matches.

Not fixing unilaterally because either direction changes user-visible score values; flagging for a maintainer decision. Aligning the local formula to the standard `1 - sim` (matching DuckDB and the "cosine similarity" score docs) seems right, plus a zero-vector guard (return NULL or max distance).

**Comment by claudespice** [comment]
### Full blast-radius analysis of the formula divergence (for the maintainer decision)

I traced every path that produces a cosine score. The **local UDF formula is the sole outlier** — every DuckDB path and the documented `_score` semantics already agree on `score = sim` (true cosine similarity):

| Path | distance | `_score = 1 - distance` |
|------|----------|--------------------------|
| Local `cosine_distance` UDF (`cosine_distance.rs`) | `(1 - sim) / 2` | `(1 + sim) / 2`  ← **outlier** |
| DuckDB federated UDF (`cosine_distance` → `array_cosine_distance`, `dialect/duckdb.rs`) | `1 - sim` | `sim` |
| DuckDB native HNSW index (`search/src/index/duckdb/metric.rs:52`) | `1 - array_cosine_distance` | `sim` |
| Snowflake | UDF denied → runs locally | same as local |

So `SELECT cosine_distance(a, b)` and every `vector_search` / `/v1/search` score differ by whether the plan federates, and only the local path disagrees with the "cosine similarity" documentation.

**Recommended direction (matches this issue's suggestion):** change the local formula from `(1 - sim) / 2` to `1 - sim` in `cosine_distance.rs`. That makes the local UDF, the federated UDF, the native HNSW index, and the docs all converge on `score = sim`.

**Why this isn't in the linked PR / why it needs a human:** flipping the formula changes user-visible `_score` values on the local path, which drifts the checked-in `_score` numbers across the entire "Test Models, AI & Search" snapshot suite (`crates/runtime/tests/models/snapshots/*_response.snap`, `*_sql_*.snap` — dozens of files). Those can only be regenerated correctly against live embedding models (`cargo insta accept` in a model-enabled run); they can't be recomputed mechanically because the stored values are `trunc(_score, 3)`. This is the maintainer call the issue flagged.

**What the PR does land now:** the separate, snapshot-safe part — the zero-vector guard. `cosine_distance` returned `Some(NaN)` for a zero-magnitude (all-zero / failed) embedding, and `NaN` sorts ahead of every real score in `ORDER BY _score DESC`, surfacing failed embeddings as top matches. The PR returns `None` (SQL `NULL`) instead, so they sort last under `NULLS LAST`. Non-zero distances are unchanged, so no snapshots drift.

---
Opened #11621 for the zero-vector guard portion.

**Comment by claudespice** [comment]
Re-checked all three bullets against current trunk (`636493775d`). **Two reproduce; the third does not and should be struck** — it was fixed after this issue was filed, deliberately and with tests.

## Bullets 1 and 2 reproduce — the formula is still `(1 - sim) / 2`

Measured against a prebuilt trunk `spiced v2.4.0-unstable-build.636493775d`, empty Spicepod, `POST /v1/sql`:

```
$ SELECT cosine_distance([1.0,0.0], [1.0,0.0])    ->  [{"identical":0.0}]
$ SELECT cosine_distance([1.0,0.0], [0.0,1.0])    ->  [{"orthogonal":0.5}]     <- 1 - sim would be 1.0
$ SELECT cosine_distance([1.0,0.0], [-1.0,0.0])   ->  [{"opposite":1.0}]       <- 1 - sim would be 2.0
```

Orthogonal at `0.5` and opposite at `1.0` is `(1 - sim) / 2` — range `[0, 1]`. The source still says so at `crates/runtime-datafusion-udfs/src/cosine_distance.rs:488`: `Some((1.0 - similarity) / 2.0)`.

The federated side is unchanged too, and does not need a live DuckDB to show — the committed search snapshots carry the rendered SQL:

```
SELECT … (1.0 - array_cosine_distance(<query vector>, "tbl"."answer_embedding")) AS "_score" …
```

(`crates/runtime/tests/search/snapshots/integration_models__search__duckdb-*.snap`). DuckDB's `array_cosine_distance` is `1 - sim`, so the federated `_score` is `sim`, while the local `_score = 1 - cosine_distance` is `(1 + sim) / 2`. Orthogonal vectors score `0.0` federated and `0.5` locally — the divergence this issue reports, still live.

## Bullet 3 does not reproduce — no NaN, and there is a guard

The issue says "no zero-vector guard — all-zero embeddings produce NaN, which sorts above all numbers in `ORDER BY _score DESC`". Both halves are now false:

```
$ SELECT cosine_distance([0.0,0.0], [1.0,0.0])                       ->  [{"zero_vec":0.5}]
$ SELECT cosine_distance([0.0,0.0], [0.0,0.0])                       ->  [{"both_zero":0.5}]
$ SELECT cosine_distance([0.0,0.0], [1.0,0.0]) IS NULL               ->  false
$ SELECT cosine_distance([sqrt(-1.0),0.0], [1.0,0.0]) IS NULL        ->  true
```

A zero-magnitude vector is answered `0.5` (treated as orthogonal), and a vector carrying a non-finite value returns **NULL**, not NaN. The code says the same at `cosine_distance.rs:473-488` — a `!is_finite()` branch returning `None`, placed deliberately *before* the zero-magnitude branch, with a comment explaining that ordering and two unit tests pinning it.

**On the instrument, because it nearly fooled me:** `/v1/sql` serializes every non-finite to JSON `null`, so a bare value read cannot tell NaN from NULL — my first pass at this looked like a fix and wasn't evidence. `IS NULL` discriminates, and it is validated in the same session: `sqrt(-1.0) IS NULL` → `false` with `isnan(sqrt(-1.0))` → `true`, so the runtime really does produce a NaN there and `IS NULL` really does distinguish it. Against that, `cosine_distance([NaN,…], …) IS NULL` → `true` means NULL, not NaN.

## What is left

Only the formula choice, which is what this issue was always parked on — and it is user-facing surface (`cosine_distance`'s returned values, and every `_score` derived from it), so per `AGENTS.md` it needs a signed-off Enhancement before the change, not a decision inside a fix PR. Note #13091 already landed against this issue as `refs`, deliberately preserving the current behaviour, so #11263 owns exactly this remaining call.

Scope: I measured the **local** path only. The federated `_score` shape above is read from committed snapshot artifacts, not from a live DuckDB federation run this session — so "orthogonal scores 0.0 federated" is inferred from that rendered SQL plus DuckDB's documented `array_cosine_distance`, and is **not** an end-to-end measurement. I also did not re-check the HNSW index path.

**Comment by claudespice** [comment]
## Industry precedent for the formula call — the naming is the outlier, not the number

Research for whoever drafts the Enhancement. Every fact below is from a page or source file fetched today; anything I could not verify is flagged rather than filled in.

**No system I could find defines a field it *calls* "cosine distance" as `(1 - sim) / 2`.** Every one uses `1 - sim` over `[0, 2]`:

| | formula | range |
|---|---|---|
| [pgvector](https://github.com/pgvector/pgvector) (`<=>`) | `1.0 - similarity`, sim clamped to `[-1,1]` (`src/vector.c`) | [0,2] |
| [DuckDB](https://duckdb.org/docs/current/sql/functions/array.html) `array_cosine_distance` | `1.0 - CosineSimilarityOp(...)` (source; the docs state no formula or range) | [0,2] |
| [Weaviate](https://docs.weaviate.io/weaviate/config-refs/distances) | `1 - cosine_sim(a,b)` | `0 <= d <= 2`, stated |
| [scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_distances.html) | "1.0 minus the cosine similarity"; impl clips to `[0.0, 2.0]` | [0,2] explicit |
| [Redis](https://redis.io/docs/latest/develop/ai/search-and-query/vectors/) | `1 - (u·v)/(‖u‖‖v‖)`, "bounded by 2" | [0,2] |
| [Azure AI Search](https://learn.microsoft.com/en-us/azure/search/vector-search-ranking) | "Cosine distance is defined as `1 - cosine_similarity`" | [0,2] |
| [SingleStore](https://docs.singlestore.com/cloud/reference/sql-reference/vector-functions/cosine-similarity-and-cosine-distance/) | "the inverse of cosine similarity" | [0,2] |
| [ClickHouse](https://clickhouse.com/docs/sql-reference/functions/distance-functions) | "one minus the cosine similarity" | range **not stated** |

**But the complement of Spice's formula has strong precedent — as a *score*, not a distance.** Spice's `_score = 1 - cosine_distance = (1 + sim) / 2`, and that is *exactly*:

- [Elasticsearch](https://www.elastic.co/docs/reference/elasticsearch/mapping-reference/dense-vector)'s `_score` for a `cosine` dense vector: `(1 + cosine(query, vector)) / 2`.
- [Weaviate](https://docs.weaviate.io/weaviate/api/graphql/additional-properties)'s `certainty`: `1 - (dist / 2)` (confirmed in `entities/additional/distance.go`), which Weaviate itself describes as "an opinionated measure that always returns a number between 0 and 1 … only usable with fixed-range distance metrics" — and which it keeps **beside** the raw `distance`, not instead of it.

So the shape Spice computes is not eccentric; **what is eccentric is calling it a distance.** Spice has an Elasticsearch/Weaviate-style bounded certainty score wearing the name every other system reserves for `1 - sim`. That reframes the choice: it is less "which formula is right" than "which of the two fields does `cosine_distance` denote", and the answer that matches every peer is `1 - sim`.

Worth noting for the score half of the decision, because the two are not the same question: peers are genuinely split on what a search API returns. [Qdrant](https://qdrant.tech/documentation/concepts/collections/) and [Milvus](https://milvus.io/docs/single-vector-search.md) return the **raw similarity** `[-1, 1]`; Elasticsearch and Weaviate `certainty` return a bounded `[0, 1]`; Azure normalizes but by reciprocal, `1 / (1 + distance)`, not by halving. Spice's [OpenAPI spec](https://spiceai.org/docs/api/http/post-search) documents `_score` as *"The similarity of the match to the query"*, which is the raw-similarity reading — and is what the DuckDB path already returns, while the local path returns `(1 + sim) / 2`. Measured locally today on trunk `636493775d`:

```
$ SELECT 1 - cosine_distance([1.0,0.0],[0.0,1.0])    ->  0.5   # orthogonal; true similarity is 0.0
$ SELECT 1 - cosine_distance([1.0,0.0],[-1.0,0.0])   ->  0.0   # opposite;   true similarity is -1.0
```

Note the second row: the local score cannot represent a negative similarity at all — opposite and orthogonal are `0.0` and `0.5`, so the bottom half of the similarity range is compressed into `[0, 0.5]`. A `score > 0.7` threshold therefore means something quite different on each path, which is the user-visible consequence this issue opened on.

**Not verified, so not relied on:** Databricks' score formula is published only as an image (alt text says "reciprocal of 1 plus the squared distance"); Snowflake ships no cosine *distance* function at all, only `VECTOR_COSINE_SIMILARITY`; Pinecone never states per-metric score semantics beyond "normalized to [-1,1]"; Qdrant's own OpenAPI describes `score` as a "distance" while its source returns a similarity; Milvus's metric page and search page contradict each other the same way. ClickHouse's range and scipy's opposite-vector value are stated nowhere I could fetch.
